modules/InboundEmail/InboundEmail.php

```diff
@@ -5401,114 +5403,138 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
 
     /**
      * Used to view non imported emails
-     * @param string $msgNo - imap mesgno
-     * @param string $uid - imap uid
+     * @param string $request - must include the metadata
      * @return Email|boolean - false on error | a non imported email
      * @throws Exception
+     * @see EmailsViewDetailNonImported::preDisplay()
      */
-    public function returnNonImportedEmail($msgNo, $uid)
+    public function returnNonImportedEmail($request)
     {
-        global $timedate, $current_user;
+        global $timedate;
+        global $current_user;
+        global $log;
+
+        $msgNo = $request['msgno'];
+        $uid = $request['uid'];
 
         if (empty($header)) {
             $email = new Email();
 
 
             $this->connectMailserver();
 
-
-            $header = imap_fetch_overview($this->conn, $uid, FT_UID);
             $fullHeader = imap_fetchheader($this->conn, $uid, FT_UID);
-            $headerByMsgNo = imap_headerinfo($this->conn, $msgNo);
-            $structure = imap_fetchstructure($this->conn, $uid, FT_UID); // map of email
-            $email->name = $this->handleMimeHeaderDecode($header[0]->subject);
+            $parsedFullHeader = imap_rfc822_parse_headers($fullHeader);
+
+            $email->name = $this->handleMimeHeaderDecode($parsedFullHeader->subject);
             $email->type = 'inbound';
 
+            if(
+                isset($request['metadata'])
+                && isset($request['metadata']['viewdefs'])
+            ) {
+                // only process field(s) in which we actually use
+                $fields_selected = $request['metadata']['viewdefs'];
+
+                if(empty($email->date_entered)) {
+                    $possibleFormats = [
+                        \DateTime::RFC2822.'+',
+                        str_replace(['D, '], '', \DateTime::RFC2822), // day-of-week is optional
+                        str_replace([':s'], '', \DateTime::RFC2822), // seconds are optional
+                        str_replace(['D, ', ':s'], '', \DateTime::RFC2822), // day-of-week is optional, seconds are optional
+                        \DateTime::RFC822,
+                        str_replace(['D, '], '', \DateTime::RFC822), // day is optional
+                        str_replace([':s'], '', \DateTime::RFC822), // seconds are optional
+                        str_replace(['D, ', ':s'], '', \DateTime::RFC822), // day is optional, seconds are optional
+                    ];
+
+                    foreach ($possibleFormats  as $possibleFormat) {
+                        $dateTime = \DateTime::createFromFormat( $possibleFormat, $parsedFullHeader->date);
+                        if ($dateTime !== false) {
```

Comment:
I don't understand this. If you can't parse the date, you just break the iteration?

Reply:
If $dateTime is a valid DateTime object we have successfully parsed a string. No need to iterate over further possible formats.

Reply:
Some IMAP server respond with different data formats. The iteration attempt to use each possible format to decode the detail. The if ($dateTime !== false) means that when the DateTime class successfully decodes the date field it will exit the loop. As we no longer need to continue trying to decode the datetime format.

I have added a comment to help explain this better.

Reply:
Unnecessary, was my mistake.